### DISCUSSION TITLE
invalid memory address or nil pointer dereference

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#3093] [Q&A]
How to fix it

```
2023/08/08 13:36:50 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
2023/08/08 13:36:50 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
2023/08/08 13:36:50 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
2023/08/08 13:36:50 RECOVERED FROM PANIC: runtime error: invalid memory address or nil pointer dereference
2023/08/08 13:36:50 goroutine 138732 [running]:
runtime/debug.Stack()
	/Users/oggy/sdk/go1.20.7/src/runtime/debug/stack.go:24 +0x65
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/tools/routine/routine.go:26 +0x5d
panic({0x15ce500, 0x2ea54b0})
	/Users/oggy/sdk/go1.20.7/src/runtime/panic.go:884 +0x213
github.com/pocketbase/pocketbase/apis.ActivityLogger.func1.1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/apis/middlewares.go:351 +0x1e9
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/tools/routine/routine.go:30 +0x8c
created by github.com/pocketbase/pocketbase/tools/routine.FireAndForget
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/tools/routine/routine.go:18 +0xac

2023/08/08 13:36:50 goroutine 138519 [running]:
runtime/debug.Stack()
	/Users/oggy/sdk/go1.20.7/src/runtime/debug/stack.go:24 +0x65
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/tools/routine/routine.go:26 +0x5d
panic({0x15ce500, 0x2ea54b0})
	/Users/oggy/sdk/go1.20.7/src/runtime/panic.go:884 +0x213
github.com/pocketbase/pocketbase/apis.ActivityLogger.func1.1.1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/apis/middlewares.go:351 +0x1e9
github.com/pocketbase/pocketbase/tools/routine.FireAndForget.func1()
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/tools/routine/routine.go:30 +0x8c
created by github.com/pocketbase/pocketbase/tools/routine.FireAndForget
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/tools/routine/routine.go:18 +0xac
```

**Comment by ganigeorgiev** [comment]
This seems to be similar to https://github.com/pocketbase/pocketbase/discussions/3067, but I'm not sure how to reproduce it.

Please provide at least some details about your setup. Are you able to reproduce it consistently? Could you share some reproducible steps?
Does it happen when the application is abruptly killed/restarted (in this case it is expected to fail but I guess the panic is not very user friendly and the message could be changed)?

**Reply by ganigeorgiev** [reply]
_I'll consider adding some guards in the next release when modifying `app.Settings()` concurrently to prevent the above panic, but in general, unless there is something else that it is triggering it, it should be safe to ignore it. You can also temporary disable the request logging by setting `0` in Admin UI > Application > Logs max days retention input (the above panic-recovery message is caused by the `ActivityLogger` middleware)._

**Reply by Oggy4th** [reply]
My code is working in production. Suddenly, it stopped and gave 404.

After downgrade pocketbase vertion to 16.9, it's not giving any error like before.

**Reply by ganigeorgiev** [reply]
Hm, this is a very strange behavior. There is not enough context about your setup to debug it but I'll try to investigate it tomorrow in more details to see if some of the v0.17.0 changes could cause it.

At the moment the only explanation for the 404 errors I can think of is if for example you've exhausted the allowed file descriptors as shown in the error in #3067 (but if that was the case you'll get the same error with the previous versions too so I don't think it is related to v0.17.0).

**Reply by Oggy4th** [reply]
I only downgraded pocketbase version. Then the error stopped.

Could you give me a list what you need, please?

**Reply by ganigeorgiev** [reply]
- Is the error consistently reproduced if you upgrade to v0.17.0? If not, then probably the same could happen with earlier versions.

- Could you provide a full error log to see if there wasn't any error before that, similar to the one in #3067?

- Do you have custom code that changes the default connection limits?

- How are you deploying PocketBase?

- Is the error reproduced locally?

- If Yes, could you create a **minimal** reproducible repo?

**Reply by ganigeorgiev** [reply]
_I've just checked the error logs for the past week on the live demo and there are no such panic-recover entries so I'm really not sure what is causing it to fail in your case. And note that the above are "handled" panic errors so usually it shouldn't disrupt your application unless there is some other error before the panics as in your previous post._

**Reply by Oggy4th** [reply]
- This error first appeared in version 16.10
- I created error.log file and invited you
- I don't think so.
- I build my code then copy the file to server.
- No, I couldn't find any way to reproduce error in locally.
- No, I can't. It's too complex too create minimal reproducible repo. Because I don't know error cause.

**Reply by ganigeorgiev** [reply]
I skimmed through the `error.log` file but couldn't find anything out of the ordinary other then some custom log messages probably manually logged during your development.

PocketBase [v0.16.10](https://github.com/pocketbase/pocketbase/releases/tag/v0.16.10) doesn't introduce any major changes that will indicate the above error (_and again, note that the above are handled panic-recover messages, so it shouldn't disrupt your application_).

I also noticed that there are a lot of restarts. Were they manually triggered or the OS terminated your program for some reason (eg. OOM)?

If not caused by an abruptly terminated process, I suspect that the error is caused by a custom code as part of your application, but I have no access to it and I can't pinpoint the exact reason or place of the error. **If you manually call somewhere `apis.ActivityLogger(app)` middleware, double check that you are passing a non-nil `app` instance (_and this `app` variable shouldn't change, eg. changing the pointer to a new PocketBase instance, as it is not concurrent safe_).**

Also when you build your application do you use CGO (_if you don't specify explicitly `CGO_ENABLED`, check the default value with `go env`_)? You can try to toggle the SQLite drivers (aka. build with `CGO_ENABLED=1 go build` and `CGO_ENABLED=0 go build`) and leave the application to run for some time with each executable and check if there would be any difference.

**Reply by ganigeorgiev** [reply]
You can also try to check your git history on the date when the error first occurred (_based on the logs it seems to be somewhere ~23/07/25_) and review your custom code for anything suspicious.

**Reply by ganigeorgiev** [reply]
I also just now noticed that in your logs you have also other panics like:

```
2023/08/08 13:36:50 http: panic serving ---: runtime error: invalid memory address or nil pointer dereference
goroutine 138789 [running]:
net/http.(*conn).serve.func1()
	/Users/oggy/sdk/go1.20.7/src/net/http/server.go:1854 +0xbf
panic({0x15ce500, 0x2ea54b0})
	/Users/oggy/sdk/go1.20.7/src/runtime/panic.go:890 +0x263
github.com/pocketbase/pocketbase/apis.LoadAuthContext.func1.1({0x2442728, 0xc004088d80})
	/Users/oggy/go/pkg/mod/github.com/pocketbase/pocketbase@v0.17.3/apis/middlewares.go:243 +0x211
github.com/labstack/echo/v5/middleware.RemoveTrailingSlashConfig.ToMiddleware.func1.1({0x2442728, 0xc004088d80})
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/middleware/slash.go:136 +0x211
github.com/labstack/echo/v5.(*Echo).ServeHTTP(0xc0007c90e0, {0x242e480, 0xc01a7e82a0}, 0xc00da94200)
	/Users/oggy/go/pkg/mod/github.com/labstack/echo/v5@v5.0.0-20230722203903-ec5b858dab61/echo.go:670 +0x234
net/http.serverHandler.ServeHTTP({0x242be60?}, {0x242e480, 0xc01a7e82a0}, 0xc00da94200)
	/Users/oggy/sdk/go1.20.7/src/net/http/server.go:2936 +0x316
net/http.(*conn).serve(0xc02da414d0, {0x242ecc0, 0xc000102870})
	/Users/oggy/sdk/go1.20.7/src/net/http/server.go:1995 +0x612
created by net/http.(*Server).Serve
	/Users/oggy/sdk/go1.20.7/src/net/http/server.go:3089 +0x5ed
```

So it is not only the `ActivityLogger`. There is definitely something else going on but I don't know how to help you debug it. For some reason in your case it seems that `app` or `app.Settings()` is nil (which should never be the case if the application was properly bootstrapped).

**Reply by Oggy4th** [reply]
thanks for your help. 

Let you know after check code again
